src/domain/mongoose_domain_sql.erl

```diff
@@ -0,0 +1,232 @@
+              LimitSQL/binary>>),
+    ok.
+
+prepare_erase() ->
+    prepare(domain_erase_settings, domain_settings, [],
+            <<"DELETE FROM domain_settings">>),
+    prepare(domain_erase_events, domain_events, [],
+            <<"DELETE FROM domain_events">>).
+
+%% ----------------------------------------------------------------------------
+%% API
+insert_domain(Domain, HostType) ->
+    transaction(fun(Pool) ->
+            case select_domain(Domain) of
+                {ok, #{host_type := HT}} when HT =:= HostType ->
+                    ok; %% ignore second call
+                {error, not_found} ->
+                    insert_domain_settings(Pool, Domain, HostType),
+                    insert_domain_event(Pool, Domain),
+                    ok;
+                {ok, _} ->
+                    {error, duplicate}
+            end
+        end).
+
+select_domain(Domain) ->
+    Pool = get_db_pool(),
+    case execute_successfully(Pool, domain_select, [Domain]) of
+        {selected, []} ->
+            {error, not_found};
+        {selected, [Row]} ->
+             {ok, row_to_map(Row)}
+    end.
+
+remove_domain(Domain, HostType) ->
+    transaction(fun(Pool) ->
+            case select_domain(Domain) of
+                {ok, #{host_type := HT}} when HT =:= HostType ->
+                    {updated, 1} = delete_domain_settings(Pool, Domain),
+                    insert_domain_event(Pool, Domain),
+                    ok;
+                {ok, _} ->
+                    {error, wrong_host_type};
+                {error, not_found} ->
+                    ok
+            end
+        end).
+
+disable_domain(Domain) ->
+    set_enabled(Domain, false).
+
+enable_domain(Domain) ->
+    set_enabled(Domain, true).
+
+erase_database() ->
+    Pool = get_db_pool(),
+    execute_successfully(Pool, domain_erase_events, []),
+    execute_successfully(Pool, domain_erase_settings, []).
+
+%% Returns smallest id first
+select_from(FromNum, Limit) ->
+    Pool = get_db_pool(),
+    Args = rdbms_queries:add_limit_arg(Limit, [FromNum]),
+    {selected, Rows} = execute_successfully(Pool, domain_select_from, Args),
+    Rows.
+
+select_updates_from(FromNum, Limit) ->
+    Pool = get_db_pool(),
+    Args = rdbms_queries:add_limit_arg(Limit, [FromNum]),
+    {selected, Rows} = execute_successfully(Pool, domain_select_events_from, Args),
+    Rows.
+
+get_min_event_id() ->
+    Pool = get_db_pool(),
+    selected_to_int(execute_successfully(Pool, domain_events_min, [])).
+
+get_max_event_id() ->
+    Pool = get_db_pool(),
+    selected_to_int(execute_successfully(Pool, domain_events_max, [])).
+
+remove_events_older_than(Id) ->
+    transaction(fun(Pool) ->
+            MaxId = get_max_event_id(),
+            if MaxId =:= 0 ->
+                    skipped;
+               MaxId < Id ->
+                    %% Removal would erase all the events, which we don't want.
+                    %% We want to keep at least one event.
```

Comment:
Why do we want to keep at least one event? At start there is no event and the server is doing fine.

Reply:
We need it to get the value of the automatic RDBMS counter. We assume that it's 0 if the table has no records. and on the very first start (before any event is added) it's indeed 0. but it's not 0 if anything has been ever added.

Reply:
But... why? If there are no records, then you don't have anything to remove, so the cleaner will just do nothing when the timeout occurs.

Reply:
how to understand then that we were disconnected for too long, and we cannot make an update using events table, e.g. we missed the events that were necessary to make an update.

Reply:
That is exactly right, we need one record, because there are some nodes, that keep LastEventId in the service gen_server. They would need to verify that their LastEventId is bigger than the minimum event id in the table (i.e. they are still allowed to sync from that table). If there is no records in the table, such check is impossible. If such check fails, the gen_server would be stopped (**TODO: node should be killed instead**).

Reply:
Ok, so there is no way to discover the issue before first insert. My guess would be to solve this by failing fast - on the first failed query to read updates. This is the time when we could ensure it loads the whole table next time.

Reply:
There is no way to discover the issue before first insert because there is no data in DB, no inserts were ever made.
And no changes were ever applied in the cluster.

Though, if there would be a node applying events normally, and someone erased event table manually (so a corner case), such node would still work, until first event appears. After that it crashes.

Reply:
I still think we are not on the same page here, I added a new comment, closing this one.